File: src/main/java/com/bajdcc/util/lexer/stringify/RegexToString.java
package com.bajdcc.util.lexer.stringify;

import com.bajdcc.util.lexer.regex.Charset;
import com.bajdcc.util.lexer.regex.Constructure;
import com.bajdcc.util.lexer.regex.IRegexComponentVisitor;
import com.bajdcc.util.lexer.regex.Repetition;

public class RegexToString implements IRegexComponentVisitor {

	/**
	 * 正则表达式的树型表达
	 */
	private StringBuilder context = new StringBuilder();

	/**
	 * 前缀
	 */
	private StringBuilder prefix = new StringBuilder();

	/**
	 * 前缀缩进
	 */
	private void appendPrefix() {
		prefix.append('\t');
		context.append(" {").append(System.lineSeparator());
	}

	/**
	 * 取消前缀缩进
	 */
	private void reducePrefix() {
		prefix.deleteCharAt(0);
		context.append(prefix).append("}").append(System.lineSeparator());
	}

	@Override
	public void visitBegin(Charset node) {
		context.append(prefix).append("字符");
		context.append(node.bReverse ? "[取反]" : "").append("\t").append(node);
		context.append(System.lineSeparator());
	}

	@Override
	public void visitBegin(Constructure node) {
		context.append(prefix).append(node.bBranch ? "分支" : "序列");
		appendPrefix();
	}

	@Override
	public void visitBegin(Repetition node) {
		context.append(prefix.toString()).append("循环{").append(node.iLowerBound).append(",").append(node.iUpperBound).append("}");
		appendPrefix();
	}

	@Override
	public void visitEnd(Charset node) {

	}

	@Override
	public void visitEnd(Constructure node) {
		reducePrefix();
	}

	@Override
	public void visitEnd(Repetition node) {
		reducePrefix();
	}

	@Override
	public String toString() {
		return context.toString();
	}
}

File: src/main/kotlin/com/bajdcc/util/lexer/stringify/RegexToString.kt
package com.bajdcc.util.lexer.stringify

import com.bajdcc.util.lexer.regex.Charset
import com.bajdcc.util.lexer.regex.Constructure
import com.bajdcc.util.lexer.regex.IRegexComponentVisitor
import com.bajdcc.util.lexer.regex.Repetition

class RegexToString : IRegexComponentVisitor {

    /**
     * 正则表达式的树型表达
     */
    private val context = StringBuilder()

    /**
     * 前缀
     */
    private val prefix = StringBuilder()

    /**
     * 前缀缩进
     */
    private fun appendPrefix() {
        prefix.append('\t')
        context.append(" {").append(System.lineSeparator())
    }

    /**
     * 取消前缀缩进
     */
    private fun reducePrefix() {
        prefix.deleteCharAt(0)
        context.append(prefix).append("}").append(System.lineSeparator())
    }

    override fun visitBegin(node: Charset) {
        context.append(prefix).append("字符")
        context.append(if (node.bReverse) "[取反]" else "").append("\t").append(node)
        context.append(System.lineSeparator())
    }

    override fun visitBegin(node: Constructure) {
        context.append(prefix).append(if (node.branch) "分支" else "序列")
        appendPrefix()
    }

    override fun visitBegin(node: Repetition) {
        context.append(prefix.toString()).append("循环{").append(node.lowerBound).append(",").append(node.upperBound).append("}")
        appendPrefix()
    }

    override fun visitEnd(node: Charset) {

    }

    override fun visitEnd(node: Constructure) {
        reducePrefix()
    }

    override fun visitEnd(node: Repetition) {
        reducePrefix()
    }

    override fun toString(): String {
        return context.toString()
    }
}
